### PR TITLE
Resources

### PR DESCRIPTION
**Pull request by spirali**

### CODE REVIEW

File: src/scheduler/scheduler.rs [hunk 1]
```diff
@@ -200,49 +215,91 @@ impl SchedulerState {
...
-        for tr in core.take_ready_to_assign().drain(..) {
-            let mut task = tr.get_mut();
-            assert!(task.is_waiting());
-            let worker_id = choose_worker_for_task(&task, core.get_worker_map(), &mut self.random);
-            log::debug!("Task {} initially assigned to {}", task.id, worker_id);
-            self.assign(core, &mut task, tr.clone(), worker_id);
+        let mut ready_tasks = core.take_ready_to_assign();
+        if !ready_tasks.is_empty() {
+            let mut tmp_workers = Vec::with_capacity(core.get_worker_map().len());
+            for tr in ready_tasks.drain(..) {
```
Comment (Kobzol) [inline src/scheduler/scheduler.rs]: `ready_tasks.into_iter()`

File: src/server/core.rs [hunk 2]
```diff
@@ -156,13 +178,34 @@ impl Core {
...
+            let task = task_ref.get();
+            if task.is_ready() {
+                self.add_ready_to_assign(task_ref.clone());
+            }
         }
         self.has_new_tasks = true;
         assert!(self.tasks.insert(task_id, task_ref).is_none());
     }
 
+    #[inline(never)]
```
Comment (Kobzol) [inline src/server/core.rs:195]: Why the inline annotations here and below?

**Review by Kobzol — APPROVED** [review]

File: src/server/core.rs (hunk 2 above)
Comment (spirali) [reply src/server/core.rs:195]: Because try_wakeup_parked_resources are called often (for every task at least once) but waking almost never happen.

Comment (Kobzol) [reply src/server/core.rs:195]: Ok. I'm not sure if it's worth it to inline a nontrivial function with a loop, but the compiler can decide anyway.

Comment (spirali) [reply src/server/core.rs:195]: It is more or less forgotten here as there were originally more complex cold/hot paths here, so probably it should be removed now.

**Pull request merged**